File: files_cpp/tests/test_order_book.cpp
#include <catch2/catch_test_macros.hpp>
#include "order_book.hpp"

using namespace depthos;

TEST_CASE("OrderBook: initial BBO is null", "[order_book]") {
    OrderBook ob("TEST_USDT");
    auto bbo = ob.bbo();
    REQUIRE(bbo == nullptr);
}

TEST_CASE("OrderBook: on_book_ticker updates BBO", "[order_book]") {
    OrderBook ob("TEST_USDT");
    std::unordered_map<std::string, std::string> data;
    data["b"] = "0.05";
    data["B"] = "100";
    data["a"] = "0.06";
    data["A"] = "200";
    auto changed = ob.on_book_ticker(data);
    REQUIRE(changed != nullptr);
    REQUIRE(changed->bid_price == Decimal("0.05"));
    REQUIRE(changed->ask_price == Decimal("0.06"));
}

TEST_CASE("OrderBook: BBO valid when bid < ask", "[order_book]") {
    OrderBook ob("TEST_USDT");
    std::unordered_map<std::string, std::string> data;
    data["b"] = "0.05";
    data["B"] = "100";
    data["a"] = "0.06";
    data["A"] = "200";
    ob.on_book_ticker(data);
    auto bbo = ob.bbo();
    REQUIRE(bbo->valid());
    REQUIRE(bbo->mid() == Decimal("0.055"));
}

File: files/tests/test_order_book.py
"""Test order book functionality."""

import pytest
from decimal import Decimal
from app.market_data.order_book import BBO, OrderBook, registry


def test_bbo_creation():
    """BBO should create correctly."""
    bbo = BBO(
        bid_price=Decimal("0.00005"),
        bid_size=100,
        ask_price=Decimal("0.00006"),
        ask_size=100,
    )
    assert bbo.bid_price == Decimal("0.00005")
    assert bbo.ask_price == Decimal("0.00006")


def test_bbo_mid():
    """BBO mid should be average of bid and ask."""
    bbo = BBO(
        bid_price=Decimal("0.00005"),
        bid_size=100,
        ask_price=Decimal("0.00006"),
        ask_size=100,
    )
    assert bbo.mid == Decimal("0.000055")


def test_bbo_spread():
    """BBO spread should be ask - bid."""
    bbo = BBO(
        bid_price=Decimal("0.00005"),
        bid_size=100,
        ask_price=Decimal("0.00006"),
        ask_size=100,
    )
    assert bbo.spread == Decimal("0.00001")


def test_order_book_registry():
    """OrderBookRegistry should create order books."""
    ob = registry.get_or_create("TEST_USDT")
    assert ob is not None
    assert ob.contract == "TEST_USDT"


def test_order_book_bbo():
    """OrderBook should track BBO."""
    ob = OrderBook("TEST_USDT")
    bbo = ob.on_book_ticker({
        "b": "0.00005",
        "B": "100",
        "a": "0.00006",
        "A": "100",
    })
    assert bbo is not None  # BBO returned
    assert ob.best_bid() == Decimal("0.00005")
    assert ob.best_ask() == Decimal("0.00006")
